### ISSUE TITLE
Optimum solution to combine/seperate/remove old and new filters and the hierarchy of them

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#10661]
I am still thinking about the optimum solution to combine/seperate/remove old and new filters and the hierarchy between them.

My current idea would be:

- Have an instance of the new filter in lists which is effective on lists only (already implemented)
- Have an instance of the new filter on map which is effective on map only (already implemented)
- Also replace the current global cache type filter by an instance of the new filter methods which then sets both of the above mentioned instances to the filtering defined here

So various use cases could be covered:
- Global filter is still in main menu but offers more options which are applied to the whole app (as users are used to even today)
- Alternatively users can setup a filter for the map only or for lists only by (instead of using the global filter). 

Would that be technically feasible and understandable/meaningful for users?

_Originally posted by @Lineflyer in https://github.com/cgeo/cgeo/issues/8383#issuecomment-842277457_

**Comment by eddiemuc** [comment]
For my personal taste the whole distinction of global vs non-global filters is already too complicated. I would simply stay with one filter saved per usage type (always the last one used for this usage type), usage types being the Cachelisttypes (map viewer, offline lists, online search of different types, map-as-list, maybe pocket queries or such)

If we really need a global filter in addition, I suggest to leave the current cache type filter as-is (maybe add multi-select). Person I would leave it out.

**Comment by Lineflyer** [comment]
> If we really need a global filter in addition, I suggest to leave the current cache type filter as-is (maybe add multi-select). Person I would leave it out.

From my support experience most users would like to have a base filter set for multiple cache types and (ideally) a limitation on D and/or T. If we could realize that in the global filter menu, that would be great for average users not interested in more complex filterings.

The question remaining would then be:
How would an active global filter interact with the filters on lower hierarchy (map/lists)?

**Comment by Lineflyer** [comment]
Suggestion for content of a potential global filter (combined with my suggestion in https://github.com/cgeo/cgeo/issues/8383#issuecomment-842324171):

- Apply to: [ ] Map&Search [ ]  Offline Lists
- Cache type (Multi select)
- D-Rating (slider from/to)
- T-Rating (slider from/to)

**Comment by eddiemuc** [comment]
> The question remaining would then be:
> How would an active global filter interact with the filters on lower hierarchy (map/lists)?

Only one option would be feasible: both filters are applied connected by AND.

If this is really wanted I can implement it. Note that I would then allow to define a fully fletched filter as global one (this is easier to implement than restricting dialog for this specific use case to type, d and t). In would probably leave out option to use OR and NOT though, since those tend to make Problems in some usage cases (e.g. on online searches)

**Comment by Lineflyer** [comment]
I tired to visualize what I had in mind and made a quick sketch out of it.
It might be, that there is still a logical problem in it or its even technically not feasible.

The idea behind is, to keep some mechanisms in place (e.g. map quick settings), have only minimal change in obvious functionality of potentially well known features, but still combine and enhance it with the new filters.

Does that make any sense?

![grafik](https://user-images.githubusercontent.com/949669/118503496-f7de9100-b72a-11eb-9b46-a487e2e819c8.png)

**Comment by eddiemuc** [comment]
> Suggestion for content of a potential global filter (combined with my suggestion in https://github.com/cgeo/cgeo/issues/8383#issuecomment-842324171):
> 
> - Apply to: [ ] Map&Search [ ]  Offline Lists
> - Cache type (Multi select)
> - D-Rating (slider from/to)
> - T-Rating (slider from/to)
> 
> 

Why make things even more complicated? If a global filter is wanted, then it is a global filter. It would then be applied simply everywhere (no selection of usage type here again). Imho this will create much confusion.

Again, I would leave out global filter completely.

**Comment by Lineflyer** [comment]
> Why make things even more complicated? If a global filter is wanted, then it is a global filter. It would then be applied simply everywhere (no selection of usage type here again). Imho this will create much confusion.

That was my proposal because a number of users will potentially not be able to understand the complexity in the new  filters. Therefore I thought an easy one level click and leave global filter covering most used options could still be helpful.

Count e.g. how many taps you need to select two cache types AND a range of D and T AND hiding found.

**Comment by eddiemuc** [comment]
> I tired to visualize what I had in mind and made a quick sketch out of it.
> It might be, that there is still a logical problem in it or its even technically not feasible.
> 
> The idea behind is, to keep some mechanisms in place (e.g. map quick settings), have only minimal change in obvious functionality of potentially well known features, but still combine and enhance it with the new filters.
> 
> Does that make any sense?
> 
> ![grafik](https://user-images.githubusercontent.com/949669/118503496-f7de9100-b72a-11eb-9b46-a487e2e819c8.png)

I am not sure I understand the diagram completely, it seems a bit complex to me.

I propose the following solution:
* One filter per usage scenario, editable wherever use case will happen. Usage types are: MapViewer, MapAsList, OfflineList, OnlineSearch, maybe PocketQuery and a few others. c:geo will simply always remember the last filter used for each usage type and apply it again next time
* Quick Settings are specific to map viewer and are only used there, nowhere else
* If global is really wanted: One global filter. If set, this filter is ADDITIVE (means: concatenated with AND condition) to ALL usage scenario. This is how global cache type filter works today too. Global filter can be restricted to use AND logic only

**Comment by Lineflyer** [comment]
Have to think it over again and will report back tonight.
Most probably we are not too far apart with our intentions, just I need to understand what filter is applied where.

**Comment by fm-sys** [comment]
What about removing the global filter and provide a setting to optionally take over the filter setting between map and lists?

**Comment by jakedot** [comment]
I would like the global filter button to remain, if not for actually filtering globally, then at least for allowing to define, store and manage user defined filters at a central location.

**Comment by Lineflyer** [comment]
OK, just a new approach of writing down my suggestion regarding the whole filtering. Forget my above picture, it was too difficult indeed.

First of all some frequent feedback I gathered from our users over the years about filtering. These are not all topics, but the main topics which appeared frequently:
1. There are many users, who had mainly two items on their wishlist:
  - Apply a filter to multiple cache types
  - Be able to filter out undesired D or T ratings
2. There are some more advanced users who would like to see more flexible ways of filtering (more than one condition, ranges of D/T, atrributes, etc.)
3. There are some users who would like to split hiding of own and found caches

Thanks to @eddiemuc item 2 is already fully covered now by the new filtering.

Item 1 is also covered by the new filtering, but I think the new filter might be already too much/complex for basic users.
That is the reason I vote to still have a global filter, but not an identical copy of the new maps/lists filter but a simplified one-level view for some most wanted basic filters. This global filter should indeed work together with the new filters in AND condition (for those who want to apply a "subfilter" on map/list level. 

Trying to describe a simple UI for such a global filter is difficult.
It should IMHO need much less taps than the advanced new filter and provide an immediate overview without any boxes to be added or menus to be opened and without knowledge about AND/OR, inversion, etc. 
I tried to make another sketch of what I see as this basic global filter set and how it could (principally) look like:
![grafik](https://user-images.githubusercontent.com/949669/118530246-b955d000-b744-11eb-8e51-9e1faf70c71f.png)
 
Once again:
The reason for proposing to keep a revised/improved global filter is to have a simple and quick filtering for the average user, who is already using the current global filter and will understand, that this is just the same with more options.
Removing it will force users to dive into the new filtering which offers more than they potentially need and has a complete new appearance and working mode (works only in parts of the app, requires different types to be added and combined, etc). Don't get me wrong: People who previously already used the old list filter will be very happy about the new perfect filtering. :) This is only about the less advanced users.

Does that make any sense to you? Otherwise we should maybe have a short meeting as this will be much faster to discuss and be understood ;)

BTW:
There is one obstacle, that needs to be decided though (see also #8383, not sure whether to discuss there or here):
The map quick settings provide hide options which are currently applied not only on map, but also on searches. My first approach was to include these into a potential new global filter. But as you can see in the sketch above I removed it again, as the working mode would be to different to the current mode (i.e.: These hide filters are currently not applied to lists for many good reasons). 
I think, that this will break the use case of many users and thus make them hard to use, because while I might be not interested in own/found/archived/disabled caches while out caching and therefore focusing on searches and maps I still want to see them in my lists (I might even have a list of all my found caches, which would then be empty once I use that global filter).
Long story short:
I would suggest to keep the same behavior as today (or come as close as possible to that), so the map quick settings for hiding caches should be "coupled" to the filter for map view and searches but **not** to lists of stored caches.
Ideally we should use this opportunity to improve those hiding filters by separating own/found (see item 3 from my initial user with list).

**Comment by eddiemuc** [comment]
I agree with the suggestion of @lineflyer. Insuggest to implement it as a switch on filter dialog "Basic" vs " Advanced" where in basic mode the filter is shown as suggested by @Lineflyer  (and this is also the default on opening it)

**Comment by Lineflyer** [comment]
> to implement it as a switch on filter dialog "Basic" vs " Advanced" where in basic mode the filter is shown as suggested by @Lineflyer (and this is also the default on opening it)

This proposal is probably even more straight forward so its the same at every level (and probably less maintenance effort in the code). We could use a well visible switch in the top bar (as in the PQ list).

What about the note to users I added in my sketch to explain that the global filter is AND to the lower level filters?
Do you it could be helpful or confusing?

**Comment by moving-bits** [comment]
Haven't been able to follow all filter-related posts throughout the day, but just want to say a big "thank you" to all of you for struggling to find the best UX for this really great new filter feature!!! ❤️

**Comment by Lineflyer** [comment]
Not to forget a big thanks to @eddiemuc for implementing this real sophisticated filtering. 🥇 I took a first deeper look on it only today.
And sorry @eddiemuc for providing more work by opening quite some issues after a bit of playing around with it ;)

**Comment by eddiemuc** [comment]
> What about the note to users I added in my sketch to explain that the global filter is AND to the lower level filters?
> Do you it could be helpful or confusing?

Hm I don't feel it to be necessary right now. The dialog should have a title like "Global filter" to make that clear, and we might add a clickable info icon somewhere where users might get more info wrt global vs "lower level".

As for the AND: I don't think this needs explanation. All the filters I can think of (including the ones on GC.com and oc.de) are pure AND filters, I think the average user expects it like this.

**Comment by MagpieFourtyTwo** [comment]
> Insuggest to implement it as a switch on filter dialog "Basic" vs " Advanced" where in basic mode the filter is shown as suggested by @Lineflyer (and this is also the default on opening it)

``Basic`` vs. ``Advanced`` switch sounds really good to me, esp. as this way the logic and the dialogue must not be implemented twice. BTW: Perhaps the "easy way" can just make use of some pre-defined saved filter sets?

This would even fit the "Global Filter" for users who are used to it, although I have never ever understood the sense of it, and still find it disturbing, esp. in it's extremely limited functionality. Why should I limit myself (even worse: globally) to e. g. Multis, missing all these >75 % Favo Tradis on my way ...? But, hey, his will is men's heaven. :)

**Comment by fm-sys** [comment]
> I just read through all the issues related to filters, because the duplicate implementation of them is something I really don't like. We have
> 
>     * global cache type filtering (directly in the main screen), just because the web site allows exactly this one type of filter up front.
> 
>     * (several) filters in the lists, which are applied to the overall contents of the list. These filters are much more flexible [...]

Just want to summarize things:
- It is confusing if two filter systems are applied in parallel
- Average users might be confused by the very flexible new filter system
- Advanced users don't like the global filter and want to see it removed

Thus, my suggestion:
- Users should be able to decide between "basic" (currently global) and "advanced" filter
- Only one of the two filter systems can be active at the same time
- If a user switches from basic to advanced, the current "basic filter" should be taken as initial value for the advanced filter (pre-configuration)
- Refactor "basic" filter, so that it uses the same code base as the advanced filter system and implement it into the filters activity
- Implement an easy accessible switch between these two modes (accessible directly within the filters activity)
- Start filters activity (in "basic" or "advanced" mode) when clicking on the filters icon at the main activity

I guess, this is the most understandable solution, which fulfills all criteria which we have aligned so far...

**Comment by eddiemuc** [comment]
Working on it...

**Comment by eddiemuc** [comment]
PR #10691 introduces basic/advanced filter config mode

**Comment by eddiemuc** [comment]
I created PR #10711 to track the necessary switch from "global" filter to the "Basic" advanced filter.

Going through the summary of @fm-sys, my understanding is that everything requested is implemented now. 
Thus I will close this issue.
